File: grid75/graphs/courseSchedule.cpp
// https://leetcode.com/problems/course-schedule/

// finding if there is a cycle in a directed graph

#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  void dfs(bool* ans, unordered_set<int>& stack, unordered_multimap<int, int>& map, int i, vector<vector<int>>& prerequisites) {
    stack.insert(i);
    while (map.find(i) != map.end()) {
      auto it = map.find(i);
      auto x = map.find(i)->second;
      map.erase(it);
      if (stack.find(x) != stack.end()) {
        *ans = false;
      } else {
        dfs(ans, stack, map, x, prerequisites);
      }
    }
    stack.erase(i);
  }
  bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    unordered_multimap<int, int> map;
    unordered_set<int> stack;
    for (int i = 0; i < prerequisites.size(); i++) {
      map.insert({prerequisites[i][0], prerequisites[i][1]});
    }
    bool ans = true;
    for (int i = 0; i < numCourses; i++) {
      dfs(&ans, stack, map, i, prerequisites);
    }
    return ans;
  }
};

File: grind75_python/courseSchedule.py
# https://leetcode.com/problems/course-schedule/

from typing import List


class Solution:
  def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

    def dfs(i):
      nonlocal ans
      stack[i] = True
      visited.add(i)
      for nei in adj_list[i]:
        if stack[nei]:
          ans = False
          return
        if not nei in visited:
          dfs(nei)
      stack[i] = False

    # make adjacency list
    adj_list = [[] for _ in range(numCourses)]
    for pair in prerequisites:
      adj_list[pair[0]].append(pair[1])

    # do dfs, check if any loops
    ans = True
    stack = [False for _ in range(numCourses)]
    visited = set()
    for i in range(len(adj_list)):
      if ans == False:
        return ans
      if (not i in visited):
        dfs(i)

    return ans


s = Solution()
print(s.canFinish(numCourses=3, prerequisites=[[0, 1], [0, 2], [1, 2]]))
